### PR TITLE
feat: Open telemetry implementation

### PR DESCRIPTION
**Pull request by surbhigarg92**
This PR adds support for [OpenTelemetry](https://opentelemetry.io/) Instrumentation for Traces and Metrics. 

Add dependency for [OpenTelemetrySDK](https://opentelemetry.io/docs/instrumentation/java/manual/#initialize-the-sdk) and required [exporters](https://opentelemetry.io/docs/instrumentation/java/exporters/).

Create OpenTelemetry object with required MeterProvider and TracerProvider exporter . Inject OpenTelemetry object via SpannerOptions or register as Global

`
OpenTelemetry openTelemetry = OpenTelemetrySdk.builder()
              .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
              .setTracerProvider(tracerProvider)
              .setMeterProvider(sdkMeterProvider)
              .build;

SpannerOptions options = SpannerOptions.newBuilder().setOpenTelemetry(openTelemetry).build();
`

 By default, OpenTelemetry traces are not enabled. To enable OpenTelemetry traces , call `SpannerOptions.enableOpenTelemetryTraces()` in startup of your application. Enabling OpenTelemetry traces will disable OpenCensus traces. Both OpenCensus and OpenTelemetry traces can not be enabled at the same time.

### CODE REVIEW

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/BatchClientImplTest.java [hunk 1]
```diff
@@ -66,6 +67,7 @@ public final class BatchClientImplTest {
   @Before
   public void setUp() {
     initMocks(this);
+    SpannerOptions.enableOpenTelemetryTraces();
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/BatchClientImplTest.java:70]: This will enable OpenTelemetry traces for all tests that come after this test. Java does not guarantee the order in which tests are executed. Are we sure that all tests will behave correctly, regardless of the order that they are executed, if this option is enabled at any random point in time?

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/SpannerRpcMetricsTest.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+  private static final Statement UPDATE_FOO_STATEMENT =
+      Statement.of("UPDATE FOO SET BAR=1 WHERE BAZ=2");
+
+  private static InMemoryMetricReader inMemoryMetricReader;
+
+  private static InMemoryMetricReader inMemoryMetricReaderInjected;
+
+  @BeforeClass
+  public static void startServer() throws IOException {
+    assumeFalse(EmulatorSpannerHelper.isUsingEmulator());
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/SpannerRpcMetricsTest.java:117]: Why is this test not supported when the emulator is used? (This is not an integration test, so it would normally not interact with the emulator)

File: README.md [hunk 3]
```diff
@@ -158,6 +158,9 @@ the Cloud Spanner Java client.
 
 ## OpenCensus Metrics
 
+> Note: OpenCensus project is deprecated. See [Sunsetting OpenCensus](https://opentelemetry.io/blog/2023/sunsetting-opencensus/).
+We recommend migrating to OpenTelemetry, the successor project.
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply README.md]: I added the complete documentation . PTAL

File: google-cloud-spanner/pom.xml [hunk 4]
```diff
@@ -16,6 +16,7 @@
   <properties>
     <site.installationModule>google-cloud-spanner</site.installationModule>
     <opencensus.version>0.31.1</opencensus.version>
+    <opentelemetry.version>1.32.0</opentelemetry.version>
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/pom.xml]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java [hunk 5]
```diff
@@ -96,152 +92,162 @@ public Timestamp writeAtLeastOnce(final Iterable<Mutation> mutations) throws Spa
...
   public ReadContext singleUse() {
-    Span span = tracer.spanBuilder(READ_ONLY_TRANSACTION).startSpan();
-    try (Scope s = tracer.withSpan(span)) {
+    ISpan span = tracer.spanBuilder(READ_ONLY_TRANSACTION);
+    try (IScope s = tracer.withSpan(span)) {
       return getSession().singleUse();
     } catch (RuntimeException e) {
-      TraceUtil.endSpanWithFailure(span, e);
+      span.setStatus(e);
+      span.end();
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java:129]: We are handling `span.end()` in AbstractReadContext [close](https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractReadContext.java#L755) method for all possible cases. So we should not add it here to avoid duplicate end on span

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/IScope.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+interface IScope extends AutoCloseable {
+  @Override
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/IScope.java:24]: Done. Please check.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java [hunk 7]
```diff
@@ -65,13 +65,20 @@ class MetricRegistryConstants {
   /** Unit to represent counts. */
   static final String COUNT = "1";
 
+  static final String Instrumentation_Scope = "cloud.google.com/java";
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/MetricRegistryConstants.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionImpl.java [hunk 8]
```diff
@@ -370,30 +370,31 @@ ApiFuture<ByteString> beginTransactionAsync(Options transactionOptions, boolean
...
-              }
-            }),
+        () -> {
+          try (IScope s = tracer.withSpan(span)) {
+            Transaction txn = requestFuture.get();
+            if (txn.getId().isEmpty()) {
+              throw newSpannerException(
+                  ErrorCode.INTERNAL, "Missing id in transaction\n" + getName());
+            }
+            span.end();
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionImpl.java:380]: No. As discussed `span.end()` are called in abstractreadcontext close call

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java [hunk 9]
```diff
@@ -55,7 +56,14 @@
...
   private static final Logger logger = Logger.getLogger(SpannerImpl.class.getName());
-  static final Tracer tracer = Tracing.getTracer();
+  final TraceWrapper tracer =
+      new TraceWrapper(
+          Tracing.getTracer(),
+          this.getOptions()
+              .getOpenTelemetry()
+              .getTracer(
+                  MetricRegistryConstants.Instrumentation_Scope,
+                  GaxProperties.getLibraryVersion(this.getOptions().getClass())));
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java:66]: Yes , verified that. If version is empty then nothing will break.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerInterceptorProvider.java [hunk 10]
```diff
@@ -29,21 +34,30 @@
...
+        new HeaderInterceptor(new SpannerRpcMetrics(GlobalOpenTelemetry.get())));
+    return new SpannerInterceptorProvider(ImmutableList.copyOf(defaultInterceptorList));
+  }
+
+  public static SpannerInterceptorProvider createDefault(OpenTelemetry openTelemetry) {
+    List<ClientInterceptor> defaultInterceptorList = new ArrayList<>();
+    defaultInterceptorList.add(new SpannerErrorInterceptor());
+    defaultInterceptorList.add(
+        new LoggingInterceptor(Logger.getLogger(GapicSpannerRpc.class.getName()), Level.FINER));
+    defaultInterceptorList.add(new HeaderInterceptor(new SpannerRpcMetrics(openTelemetry)));
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerInterceptorProvider.java:53]: Yes, null handling is there

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/OpenCensusSpan.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+  @Override
+  public void addAnnotation(String message, Map<String, Object> attributes) {
+    Map<String, AttributeValue> ocAttributeValues = new HashMap<>();
+    for (Map.Entry<String, Object> entry : attributes.entrySet()) {
+      String key = entry.getKey();
+      Object val = entry.getValue();
+      if (val == null || val instanceof String) {
+        String strVal = (String) val;
+        assert strVal != null;
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/OpenCensusSpan.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/OpenCensusSpan.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    }
+    return ImmutableMap.of();
+  }
+
+  @Override
+  public void addAnnotation(String message, Map<String, Object> attributes) {
+    Map<String, AttributeValue> ocAttributeValues = new HashMap<>();
+    for (Map.Entry<String, Object> entry : attributes.entrySet()) {
+      String key = entry.getKey();
+      Object val = entry.getValue();
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/OpenCensusSpan.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java [hunk 13]
```diff
@@ -1312,6 +1333,55 @@ public static void useDefaultEnvironment() {
     SpannerOptions.environment = SpannerEnvironmentImpl.INSTANCE;
   }
 
+  /**
+   * Enables OpenTelemetry traces. Enabling OpenTelemetry traces will disable OpenCensus traces. By
+   * default, OpenCensus traces are enabled.
+   */
+  public static void enableOpenTelemetryTraces() {
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java:1345]: This documentation is added in SetOpenTelemetry. Also added the details the logic to handle that enableOpenTelemetryTraces should be called before Spanner client gets created

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java [hunk 14]
```diff
@@ -140,6 +146,14 @@ public class SpannerOptions extends ServiceOptions<Spanner, SpannerOptions> {
...
   private final boolean attemptDirectPath;
   private final DirectedReadOptions directedReadOptions;
+  private final OpenTelemetry openTelemetry;
+
+  enum TracingFramework {
+    OPEN_CENSUS,
+    OPEN_TELEMETRY
+  }
+
+  private static TracingFramework activeTracingFramework = TracingFramework.OPEN_CENSUS;
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java]: As discussed added the lock to set this.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/TraceWrapper.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+  private final Tracer openCensusTracer;
+  private final io.opentelemetry.api.trace.Tracer openTelemetryTracer;
+
+  TraceWrapper(Tracer openCensusTracer, io.opentelemetry.api.trace.Tracer openTelemetryTracer) {
+    this.openTelemetryTracer = openTelemetryTracer;
+    this.openCensusTracer = openCensusTracer;
+  }
+
+  ISpan spanBuilder(String spanName) {
+    if (SpannerOptions.getActiveTracingFramework().equals(TracingFramework.OPEN_TELEMETRY)) {
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/TraceWrapper.java:36]: As discussed we now allow this to be set only once, If a user tries to reset it, they will get error

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/TraceWrapper.java [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+      return new OpenTelemetrySpan(openTelemetryTracer.spanBuilder(spanName).startSpan());
+    } else {
+      return new OpenCensusSpan(openCensusTracer.spanBuilder(spanName).startSpan());
+    }
+  }
+
+  ISpan spanBuilderWithExplicitParent(String spanName, ISpan parentSpan) {
+    if (SpannerOptions.getActiveTracingFramework().equals(TracingFramework.OPEN_TELEMETRY)) {
+      OpenTelemetrySpan otParentSpan;
+      if (!(parentSpan instanceof OpenTelemetrySpan)) {
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/TraceWrapper.java]: Removed this check

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerRpcViews.java [hunk 17]
```diff
@@ -91,28 +93,49 @@ public class SpannerRpcViews {
...
+   * @deprecated The OpenCensus project is deprecated. Use OpenTelemetry to get gfe_latency and
+   *     gfe_header_missing_count metrics.
    */
   @VisibleForTesting
+  @ObsoleteApi(
+      "The OpenCensus project is deprecated. Use OpenTelemetry to get gfe_latency and gfe_header_missing_count metrics.")
   public static void registerGfeLatencyAndHeaderMissingCountViews() {
-    viewManager.registerView(SPANNER_GFE_LATENCY_VIEW);
-    viewManager.registerView(SPANNER_GFE_HEADER_MISSING_COUNT_VIEW);
+    if (SpannerOptions.isEnabledOpenCensusMetrics()) {
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerRpcViews.java:104]: Metrics its little different. OpenCensus Metrics and OpenTelmetry metrics can be enabled or disable separately. So enabling one will not interfere with other. Also customers don't have the option to disable OpenTelemetry metrics . 

If customer creates 1 client and enables metrics and creates another spanner client. Customer would only get metrics for the second client which is expected.

File: README.md [hunk 18]
```diff
@@ -156,55 +156,127 @@ See [Session Pool and Channel Pool Configuration](session-and-channel-pool-confi
...
+<dependency>
+    <groupId>io.opentelemetry</groupId>
+    <artifactId>opentelemetry-exporter-otlp</artifactId>
+    <version>{opentelemetry.version}</version>
+</dependency>
+```
+If you are using Gradle, add this to your dependencies
+```Groovy
+compile 'io.opencensus:opencensus-impl:0.30.0'
+compile 'io.opencensus:opencensus-exporter-stats-stackdriver:0.30.0'
```
Comment (olavloite) [inline README.md:220]: Shouldn't this be:

```suggestion
compile 'io.opentelemetry:opentelemetry-sdk:{opentelemetry.version}'
compile 'io.opentelemetry:opentelemetry-sdk-metrics:{opentelemetry.version}'
compile 'io.opentelemetry:opentelemetry-exporter-oltp:{opentelemetry.version}'
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/OpenCensusSpan.java [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+import io.opencensus.trace.EndSpanOptions;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.Status;
+import java.util.HashMap;
+import java.util.Map;
+
+public class OpenCensusSpan implements ISpan {
+
+  static final EndSpanOptions END_SPAN_OPTIONS =
+      EndSpanOptions.builder().setSampleToLocalSpanStore(true).build();
```
[the remark this answers is not in the document]
Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/OpenCensusSpan.java:31]: Ack

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java [hunk 20]
```diff
@@ -1312,6 +1345,77 @@ public static void useDefaultEnvironment() {
...
+        throw new IllegalStateException(
+            "ActiveTracingFramework is set to OpenTelemetry and cannot be reset after SpannerOptions object is created.");
+      }
+      activeTracingFramework = TracingFramework.OPEN_CENSUS;
+    }
+  }
+
+  @ObsoleteApi(
+      "The OpenCensus project is deprecated. Use enableOpenTelemetryTraces to switch to OpenTelemetry traces")
+  /**
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java:1379]: nit: The Javadoc comment should be placed before the `@ObsoleteApi` annotation

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java [hunk 21]
```diff
@@ -1312,6 +1345,77 @@ public static void useDefaultEnvironment() {
...
+  }
+
+  /** Enables OpenTelemetry metrics. Enable OpenTelemetry metrics before creating Spanner client. */
+  public static void enableOpenTelemetryMetrics() {
+    SpannerOptions.enableOpenTelemetryMetrics = true;
+  }
+
+  public static boolean isEnabledOpenTelemetryMetrics() {
+    return SpannerOptions.enableOpenTelemetryMetrics;
+  }
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java:1417]: We can do it in a separate PR, but I think that these static options should also be guarded by the same lock as the one that we use for `activeTracingFramework`.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerRpcViews.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerRpcViews.java:104]: Hmm.... I think it will depend on what is called in which order. But as this is OpenCensus, which only supports global configuration, I guess this is the best we can do.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/BatchClientImplTest.java (hunk 1 above)
Comment (olavloite) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/BatchClientImplTest.java:71]: As it is a setting that you want to be enabled globally for all the tests in this class, you should put it in a class setup method. So like this:

```java
@BeforeClass
public static void setupOpenTelemetry() {
  SpannerOptions.enableOpenTelemetryTraces();
}
```

And put the reset in a cleanup method that is called after all the tests in this class have finished.

```java
@AfterClass
public static void resetOpenTelemetry() {
  SpannerOptions.resetActiveTracingFramework();
}
```

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/OpenTelemetrySpanTest.java [hunk 22]
```diff
@@ -0,0 +1,561 @@
...
+      Assume.assumeTrue(
+          "Skipping test as reflection is not allowed on reflection class in this JDK build",
+          false);
+    }
+    modifiersField.setAccessible(true);
+    // Remove the final modifier from the 'traceComponent' field.
+    modifiersField.setInt(field, field.getModifiers() & ~Modifier.FINAL);
+    field.set(null, failOnOverkillTraceComponent);
+
+    // OpenTelemetry Configuration
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/OpenTelemetrySpanTest.java:234]: nit: remove comment, or add configuration :-)

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ResumableStreamIteratorTest.java [hunk 23]
```diff
@@ -163,12 +170,39 @@ public void simple() {
   }
 
   @Test
-  public void closedSpan() {
+  public void closedOTSpan() {
+    SpannerOptions.resetActiveTracingFramework();
+    SpannerOptions.enableOpenTelemetryTraces();
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/ResumableStreamIteratorTest.java:175]: Are you sure you need this in a test? If so, then you should do it like this:

```
try {
  SpannerOptions.enableOpenTelemetryTraces();
} finally {
  SpannerOptions.resetActiveTracingFramework()
}
```

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ResumableStreamIteratorTest.java [hunk 24]
```diff
@@ -163,12 +170,39 @@ public void simple() {
...
+    assertThat(consume(resumableStreamIterator)).containsExactly("a", "b").inOrder();
+
+    resumableStreamIterator.close("closed");
+    verify(oTspan).end();
+  }
+
+  @Test
+  public void closedOCSpan() {
+    SpannerOptions.resetActiveTracingFramework();
+    SpannerOptions.enableOpenCensusTraces();
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/ResumableStreamIteratorTest.java:200]: Here also: We should prefer not to do this kind of set/reset in a separate test, and if we do, then it should preferably be in the other order:
1. Set the desired configuration
2. Reset to the default in a finally block

The above will ensure that we don't unintentionally leak a custom configuration to other tests. Now, OpenCensus traces are enabled for any test that follows this test. That can give unintended side effects, as much of this configuration is global.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java [hunk 25]
```diff
@@ -1757,6 +1831,190 @@ public void testSessionMetrics() throws Exception {
...
+    setupMockSessionCreation();
+    pool = createPool(clock, metricRegistry, labelValues);
+    PooledSessionFuture session1 = pool.getSession();
+    PooledSessionFuture session2 = pool.getSession();
+    session1.get();
+    session2.get();
+
+    MetricsRecord record = metricRegistry.pollRecord();
+    assertThat(record.getMetrics().size()).isEqualTo(0);
+    SpannerOptions.enableOpenCensusMetrics();
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java:1864]: Put in a finally block

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java [hunk 26]
```diff
@@ -1757,6 +1831,190 @@ public void testSessionMetrics() throws Exception {
...
+  }
+
+  @Test
+  public void testOpenTelemetrySessionMetrics() throws Exception {
+    SpannerOptions.enableOpenTelemetryMetrics();
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java:1869]: Reset this in a finally block (and check other places as well)

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpanTest.java [hunk 27]
```diff
@@ -154,11 +168,30 @@ public static void stopServer() throws InterruptedException {
 
   @Before
   public void setUp() throws Exception {
+    SpannerOptions.resetActiveTracingFramework();
+    SpannerOptions.enableOpenCensusTraces();
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SpanTest.java:172]: 1. Put the enable... in a @BeforeClass method
2. Put the reset... in a @AfterClass method

This will ensure that the configuration of this test class does not leak into any other tests.

**Review by olavloite — APPROVED** [review]
LGTM, with some nits on the tests. Especially the configuration set/reset strategy should be changed for the tests, so that a test always starts with setting its own desired configuration, and then at the end resets to the default. That will ensure that the configuration for one specific test does not unintentionally leak to another test.
